### PR TITLE
Make headless napari usable without qt

### PR DESCRIPTION
**Pull request by tlambert03**
# Description
This WIP PR allows the majority of non-GUI (headless) napari functionality work without Qt even being installed:
```
pip uninstall qtpy
pytest --ignore napari/_qt --ignore napari/_tests --ignore napari/_vispy
```
with that command, the only thing that fails is `napari/plugins/_tests/test_reader_plugins.py` ... which can be changed in this PR.  Otherwise, all tests in components, layers, utils, and plugins are working (as long as I remove `gui_qt` from `napari.__init__`.  We should have a chat about `gui_qt` here as well.  It'd be nice if that could be wrapped as `napari.event_loop` or something.  I recognize that's a major API change and would need to be deprecated... but following some recent discussions, we may want to move away from that pattern and towards something like instantiating the QtApp on demand and calling `viewer.run()` to block anyway.

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by sofroniewn** [comment]
> This WIP PR allows the majority of non-GUI (headless) napari functionality work without Qt even being installed:

I'm very excited about this @tlambert03, thanks for picking it up! I see things like this as paving the way for us to maybe one day separate the repo into `napari-models` and `napari`, where `napari-models` would contain all our model code that is Qt/ vispy independent (and maybe fully pydanticlly validated!) and `napari` would contain the current vispy/ Qt front end. If we ever had another front-end they could still use `napari-models` and would know exactly what they were getting!! Are you thinking along these lines too? (Note I don't think we're ready for that now, but when we are it would be amazing)

> We should have a chat about gui_qt here as well. It'd be nice if that could be wrapped as napari.event_loop or something. I recognize that's a major API change and would need to be deprecated... but following some recent discussions, we may want to move away from that pattern and towards something like instantiating the QtApp on demand and calling viewer.run() to block anyway.

I'm open to these ideas too, but if we're going to be making some major breaking changes here though I would like us to at least briefly reconsider some of the ideas in https://github.com/napari/napari/pull/1799#issuecomment-720002919 where we'd move the window off the Viewer and provide the `Window` through a different API - note the goal isn't to prevent users from accessing Window or Qt stuff, but to make it slightly more explicit when they do. I will add that if we create that second `napari-models` repo then maybe the distinction between `ViewerModel` and `Viewer` actually may become more clear to people and it's might be ok to leave window on this new Viewer in the napari repo, but I'm still not sure

**Comment by tlambert03** [comment]
ok this is ready for review and discussion.  All it really amounts to is moving the qt-specific and vispy-specific `__init__` stuff  into `_qt.__init__` and `_vispy.__init__` respectively.  And moving the `build_resources` logic and stylesheets into `_qt/resources` (since that's all qt-specific stuff).

2 main things to discuss:
- The _only_ thing preventing importing and using napari without Qt even being installed at this point is just the `from ._qt import gui_qt` line in `napari.__init__`.  I could create a wrapper at `napari.utils.event_loop` that tries to import from `_qt`?
- Similarly I've [modified the  `from ._qt import Window` in `napari/viewer.py` ](https://github.com/napari/napari/pull/2039/files#diff-d5c7eaa497f40561af9dc89131f265676417b726c6eb454b11281e0cbe25c3f2R50-R64) to create a mock Window object that will only raise an exception if someone tries to access anything on `viewer.window`.  Could move that to a dedicated file, or modify as requested.

**Comment by tlambert03** [comment]
> Are you thinking along these lines too? (Note I don't think we're ready for that now, but when we are it would be amazing)

yep. clear division between front-end is a goal here.  we can probably speed up import times for non-gui ipython usage if we delay importing vispy and qt until the viewer actually needs to be shown.

> I'm open to these ideas too, but if we're going to be making some major breaking changes here though I would like us to at least briefly reconsider some of the ideas in #1799 (comment) where we'd move the window off the Viewer and provide the Window through a different API - note the goal isn't to prevent users from accessing Window or Qt stuff, but to make it slightly more explicit when they do.

I don't want to make any breaking changes here.  The stuff I alluded to there can be done in another PR.  I still see no problem with `viewer.window` being right there.  Let's have that conversation elsewhere

File: napari/_qt/resources/_tests/test_resources.py
Comment (sofroniewn) [inline napari/_qt/resources/_tests/test_resources.py]: Don't we generally do absolute imports in our tests?

File: napari/_qt/resources/build_icons.py
Comment (sofroniewn) [inline napari/_qt/resources/build_icons.py]: Is there a reason this an absolute import and not relative? It looks like the theme import is still relative

File: napari/_vispy/__init__.py
Comment (sofroniewn) [inline napari/_vispy/__init__.py:6]: Makes a lot of sense to have this here. This will always happen after the `_qt` import happens in our code base with current usage, but I'm trying to think if that is necessary - i.e. would we need to put `_vispy` inside `_qt`, I don't think so, but just asking in case I'm missing anything

File: napari/__init__.py
Comment (sofroniewn) [inline napari/__init__.py:20]: I guess the `_magicgui` import no longer requires any qt imports which is nice, right? Or does it still do some.

File: napari/__init__.py
Comment (sofroniewn) [inline napari/__init__.py:16]: Ok, I see this is the place where everything qt gets imported and you're wondering about doing the "napari.utils.event_loop" import? I'm open to that, but I'm not sure, I really understand how that will work. I might need to think on it more. Could come after this PR merges

Comment (sofroniewn) [reply napari/__init__.py:16]: Actually doesn't all the Qt stuff get imported when the Viewer gets imported above during the Window import?

File: napari/_qt/__init__.py
Comment (sofroniewn) [inline napari/_qt/__init__.py:6]: Does moving this here have any unexpected consequences for out "public" qt folder? I guess not as first thing there is `from .._qt.qt_main_window import Window` which will trigger this code, but I just want to check

File: napari/viewer.py
Comment (sofroniewn) [inline napari/viewer.py]: > Similarly I've modified the from ._qt import Window in napari/viewer.py to create a mock Window object that will only raise an exception if someone tries to access anything on viewer.window. Could move that to a dedicated file, or modify as requested.

I think I'd rather have this in a dedicated file to try and keep `Viewer` as clean and simple as possible.

**Review by sofroniewn — APPROVED** [review]
I'm approving this, I've left a whole bunch of questions/ comments - most of which are to make sure I understand everything.

Overall I love how the Qt initialization is now in the Qt init and the vispy initialization is in the vispy init, those two changes alone means this smells right to me.

> I don't want to make any breaking changes here. The stuff I alluded to there can be done in another PR. I still see no problem with viewer.window being right there. Let's have that conversation elsewhere

Yup, that's fair enough, as long as we aren't making breaking changes no need to worry about that other stuff yet.

> The only thing preventing importing and using napari without Qt even being installed at this point is just the from ._qt import gui_qt line in napari.__init__. I could create a wrapper at napari.utils.event_loop that tries to import from _qt?

I'm open to this, but I wonder if we just get this PR merged without it as it's already doing a lot of good, and then add that in. Might be easier for me to parse broken up

> Similarly I've modified the from ._qt import Window in napari/viewer.py to create a mock Window object that will only raise an exception if someone tries to access anything on viewer.window. Could move that to a dedicated file, or modify as requested.

As noted in my comment above I'd prefer this to be in a dedicated file.

I also wonder if/ how we should force code to go down this path if qt is imported but a person doesn't want to use them yet. We can save that convo for the future. I'd rather not add more kwargs/ env variables right now, but just something to think about.

Overall nice work @tlambert03!!

File: napari/_qt/resources/_tests/test_resources.py
Comment (tlambert03) [reply napari/_qt/resources/_tests/test_resources.py]: oh right, sure can change

File: napari/_qt/resources/build_icons.py
Comment (tlambert03) [reply napari/_qt/resources/build_icons.py]: well the "reason" was that i felt like it was clearer to put `import napari.resources` than to say `from ... import resources`... (particularly since `from .. import resources` is _also_ a real import ...)  
but i get that it's not strictly in keeping with the relative imports we use.  so let me know if you want it changed

File: napari/_vispy/__init__.py
Comment (tlambert03) [reply napari/_vispy/__init__.py:6]: honestly, you can remove the whole `from vispy import app` and it's still fine.  I just left it in there cause it was in `napari.__init__`.  But vispy [tries these core backends](https://github.com/vispy/vispy/blob/7f3f6f60c8462bb8a3a8fa03344a2e6990b86eb2/vispy/app/backends/__init__.py#L15-L27) (defaulting to the Qt backends first).  So the result is the same either way.  It's conceivable that there is some combination of environment that might mess this up (but I think it would also be messed in master)

File: napari/_qt/resources/build_icons.py
Comment (sofroniewn) [reply napari/_qt/resources/build_icons.py]: lol, we can bump this one to @jni. I don't have strong feelings, was just thinking about consistency, but I might say that having both `from ... import resources` and `from .. import resources` be valid suggests we might want to rename one of them! We could take a cue from how we deal with utils and go with `qt_resources` for the inner one. (In which case this would still be `from ... import resources`). We don't always do that with utils, but might even want to too.

I'll let @jni weigh in though too, he might have a better idea!

File: napari/__init__.py
Comment (tlambert03) [reply napari/__init__.py:16]: With the current PR, all the Qt stuff gets imported when a Viewer is _instantiated_ (not just when it's imported) because  `from ._qt import Window` is in the init.  I'm happy to move that to a new file, but I think keeping some `get_window_class()` helper inside the init, thereby keeping the whole Qt chain import lazy is maybe a nice thing to do?

File: napari/_vispy/__init__.py
Comment (sofroniewn) [reply napari/_vispy/__init__.py:6]: Interesting, I think we might have needed it a long time ago, but if we don't now could drop, or could just leave in to be explicit. I don't mind, but am fine with it staying. Your call if you want to drop

File: napari/__init__.py
Comment (tlambert03) [reply napari/__init__.py:20]: correct, no more qt imports

File: napari/__init__.py
Comment (sofroniewn) [reply napari/__init__.py:16]: ah right right, i get it now. Yes, let's keep Qt chain import lazy. A helper function sounds nice.

File: napari/_qt/__init__.py
Comment (tlambert03) [reply napari/_qt/__init__.py:6]: yeah, don't think so

File: napari/viewer.py
Comment (tlambert03) [reply napari/viewer.py]: i create a `components.window` module. let me know how that strikes you.  I think it could be a place, in the future, where we define what a graphical backend needs to implement to work with napari.  (something like the [widget protocol in magicgui](https://github.com/napari/magicgui/blob/master/magicgui/widgets/_protocols.py))

File: napari/_vispy/__init__.py
Comment (tlambert03) [reply napari/_vispy/__init__.py:6]: eh, I don't mind it there.  It reminds us that `vispy.app.use_app` exists.

File: napari/_qt/resources/build_icons.py
Comment (tlambert03) [reply napari/_qt/resources/build_icons.py]: renamed `_qt.resources` to `_qt.qt_resources`

File: napari/viewer.py
Comment (sofroniewn) [reply napari/viewer.py]: File looks good and I like the idea, but I think i'd rather it go at the top level instead of in `components`, I don't really like having to go `from .._qt import Window` inside `components` and could one day see the whole components move to a `napari-models` folder which would be entirely independent from all our backends. Thoughts?

Comment (tlambert03) [reply napari/viewer.py]: top level as in `napari.window`?  that's fine with me too

Comment (sofroniewn) [reply napari/viewer.py]: An alternative view would be that the `components` folder would stay inside `napari`, there would be no `napari-models`, but instead all the qt/ vispy stuff would go to `napari-qt` and then we'd have `napari-electron` or something else, that you could have behind `napari`. We don't need to decide this now, and we probably want a top level window import regardless ...

Comment (sofroniewn) [reply napari/viewer.py]: yeah, as I just noted, i can see arguments for the other direction too, but I'd probably go with `napari.window` for now

**Comment by tlambert03** [comment]
also was able to use `ViewerModel()` instead of `make_test_viewer` to test reader plugins.  So all of our plugin tests now also run without qtpy installed

**Comment by sofroniewn** [comment]
I love this! I'm wondering could we add a test that would install napari without any Qt and then run the subset of tests that we think should pass without it? Could ensure that we don't have accidental regressions here where we mix qt things back into our non-qt part of the codebase.

```
pip install napari
pytest --ignore napari/_qt --ignore napari/_tests --ignore napari/_vispy
```

Could even check no Qt present too!!!

**Comment by tlambert03** [comment]
> I love this! I'm wondering could we add a test that would install napari without any Qt and then run the subset of tests that we think should pass without it? Could ensure that we don't have accidental regressions here where we mix qt things back into our non-qt part of the codebase.

I like it too! :) I was thinking the same thing... however, each time I run all those tests, I have to comment out `from ._qt import gui_qt` in `napari/__init__.py`, so maybe let's talk about that little lingering bit there in a follow up PR, and then try to add a test to prevent a regression here?

for the `gui_qt` the least invasive thing I was thinking is to have a more generically named `napari.event_loop` context manager, and for now (and many releases from now) have a `napari.gui_qt = napari.event_loop` alias that lets us gradually move away from backend specific language without breaking anything.  Anyway... will open a new issue for that

**Comment by sofroniewn** [comment]
Ah yes, ok. Let's get this now as is, and then try and pick up the rest in a follow up PR, including the event loop stuff, which I'm broadly supportive of too.

From my perspective this is now done. It is a backwards compatible change, but a fairly large one so I want to let it sit for a little in case @jni has additional comments and then merge, unless there is anything else that you think needs to be added @tlambert03

File: napari/_qt/qt_resources/__init__.py
Comment (jni) [inline napari/_qt/qt_resources/__init__.py:42]: Won't the parent dirs be left lying around by this function? That seems suboptimal...

File: napari/window.py
Comment (jni) [inline napari/window.py:6]: Love this! Is there a reason for the except clause and dummy window right now?

**Comment by jni** [comment]
Thank you @tlambert03 and @sofroniewn! It's been nice to see all these changes come through while I was chilling out. I've left a couple of questions mostly for my own enlightenment, but I really love these changes. The lazy importing of Qt also means that `napari --info` will have a much better chance of always working — which is awesome!

File: napari/_qt/qt_resources/__init__.py
Comment (tlambert03) [reply napari/_qt/qt_resources/__init__.py:42]: `target.unlink()` below is supposed to clean it up (though I see that depending on the nature of some exception it might not).  But that code is many months old (just moved from `napari.resources.__init__` and I haven't seen anything lying around yet

File: napari/window.py
Comment (tlambert03) [reply napari/window.py:6]: The main reason for the dummy window is providing a no-op `close()` function that would otherwise raise an unnecessary exception upon viewer.close() [here](https://github.com/tlambert03/napari/blob/676421a22ee2e8d02e922d7a657b1705d517c2f5/napari/viewer.py#L104)

**Pull request merged**
